### PR TITLE
Extract and reuse common sky/shell components on Mac and iOS

### PR DESCRIPTION
**Pull request by chinmaygarde**
The Mac target is a work in progress. I am also in the process of extracting the test runner methods into a common file. In a separate commit.

### CODE REVIEW

File: sky/shell/ios/main_ios.mm [hunk 1]
```diff
@@ -22,49 +17,9 @@ unsigned long getauxval(unsigned long type) {
...
-  InitializeLogging();
-  result = base::i18n::InitializeICU();
-  DLOG_ASSERT(result);
-  result = gfx::GLSurface::InitializeOneOff();
-  DLOG_ASSERT(result);
-  return UIApplicationMain(argc, argv, nil,
-                           NSStringFromClass([SkyAppDelegate class]));
+int main(int argc, const char * argv[]) {
+  return PlatformMacMain(argc, argv, ^(){
+    return UIApplicationMain(argc, (char **)argv, nil,
```
Comment (eseidelGoogle) [inline sky/shell/ios/main_ios.mm]: I'm surprised you need this cast here.  Is main's arg actually const?

File: sky/shell/ios/main_ios.mm [hunk 2]
```diff
@@ -22,49 +17,9 @@ unsigned long getauxval(unsigned long type) {
...
+                             NSStringFromClass([SkyAppDelegate class]));
```
Comment (eseidelGoogle) [inline sky/shell/ios/main_ios.mm]: The class doesn't have a method for this?  SkyAppDelegate.class.name?

File: sky/shell/mac/Info.plist [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+	<key>CFBundleShortVersionString</key>
+	<string>1.0</string>
+	<key>CFBundleSignature</key>
+	<string>????</string>
+	<key>CFBundleVersion</key>
+	<string>1</string>
+	<key>LSMinimumSystemVersion</key>
+	<string>10.6</string>
+	<key>NSHumanReadableCopyright</key>
+	<string>Copyright © 2015 The Chromium Authors. All rights reserved.</string>
```
Comment (eseidelGoogle) [inline sky/shell/mac/Info.plist]: Are we still the Chromium authors or now the Mojo authors?

File: sky/shell/mac/main_mac.mm [hunk 4]
```diff
@@ -0,0 +1,83 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#import <Cocoa/Cocoa.h>
+#include <iostream>
```
Comment (eseidelGoogle) [inline sky/shell/mac/main_mac.mm]: Do you need iostream?

File: sky/shell/mac/main_mac.mm [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+#include "base/bind.h"
+#include "base/message_loop/message_loop.h"
+#include "sky/engine/public/web/WebRuntimeFeatures.h"
+#include "sky/shell/mac/platform_mac.h"
+#include "base/command_line.h"
```
Comment (eseidelGoogle) [inline sky/shell/mac/main_mac.mm]: Sort plz. :)

File: sky/shell/mac/main_mac.mm [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+      (!command_line.HasSwitch(sky::shell::switches::kPackageRoot) &&
+       !command_line.HasSwitch(sky::shell::switches::kSnapshot))) {
+    return false;
+  }
+
+  return true;
+}
+
+void Usage() {
+  std::cerr << "(For Test Shell) Usage: sky_shell"
```
Comment (eseidelGoogle) [inline sky/shell/mac/main_mac.mm]: What's the (For Test Shell)?

File: sky/shell/mac/main_mac.mm [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+            << " --" << switches::kNonInteractive
+            << " --" << switches::kPackageRoot << "=PACKAGE_ROOT"
+            << " --" << switches::kSnapshot << "=SNAPSHOT"
+            << " [ MAIN_DART ]" << std::endl;
+}
+
+void Init() {
+  base::CommandLine& command_line = *base::CommandLine::ForCurrentProcess();
```
Comment (eseidelGoogle) [inline sky/shell/mac/main_mac.mm]: Init probably could be shared between all these in a later patch.

File: sky/shell/mac/platform_mac.mm [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+#include "sky/shell/service_provider.h"
+#include "sky/shell/ui_delegate.h"
+#include "base/lazy_instance.h"
+#include "base/message_loop/message_loop.h"
+
+static void InitializeLogging() {
+  logging::LoggingSettings settings;
+  settings.logging_dest = logging::LOG_TO_SYSTEM_DEBUG_LOG;
+  logging::InitLogging(settings);
+  logging::SetLogItems(false,   // Process ID
```
Comment (eseidelGoogle) [inline sky/shell/mac/platform_mac.mm]: And you see why we've banned raw bools as arguments in Blink/Sky code. :)

File: sky/shell/mac/sky_app_delegate.m [hunk 9]
```diff
@@ -0,0 +1,15 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#import "sky_app_delegate.h"
+
+@interface SkyAppDelegate ()
```
Comment (eseidelGoogle) [inline sky/shell/mac/sky_app_delegate.m]: Do you want to be an NSApplicationDelegate?

File: sky/shell/mac/sky_window.h [hunk 10]
```diff
@@ -0,0 +1,9 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#import <Cocoa/Cocoa.h>
+
+@interface SkyWindow : NSWindow
```
Comment (eseidelGoogle) [inline sky/shell/mac/sky_window.h]: This seems very atypical.  Normally you subclass NSView and not NSWindow.

File: sky/shell/mac/sky_window.mm [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+  NSAssert(_shell_view == nullptr, @"The shell view must not already be set");
+  auto shell_view = new sky::shell::ShellView(sky::shell::Shell::Shared());
+  _shell_view.reset(shell_view);
+
+  auto widget = reinterpret_cast<gfx::AcceleratedWidget>(self.renderSurface);
+  self.platformView->SurfaceCreated(widget);
+}
+
+-(NSString *) skyInitialLoadURL {
+  return @"http://localhost:8080/sky/sdk/example/rendering/simple_autolayout.dart";
```
Comment (eseidelGoogle) [inline sky/shell/mac/sky_window.mm]: Do you mean for this?

File: sky/shell/mac/sky_window.mm [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+
+  // sky::ViewportMetricsPtr metrics = sky::ViewportMetrics::New();
+  // metrics->physical_width = size.width * scale;
+  // metrics->physical_height = size.height * scale;
+  // metrics->device_pixel_ratio = scale;
+  // _sky_engine->OnViewportMetricsChanged(metrics.Pass());
+}
+
+-(void) setupSurfaceIfNecessary {
+  if (self.isSurfaceSetup) {
```
Comment (eseidelGoogle) [inline sky/shell/mac/sky_window.mm]: Did you mean for the { }

File: sky/shell/mac/sky_window.mm [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+    return;
+  }
+
+  self.surfaceSetup = YES;
+
+  [self setupShell];
+  [self setupAndLoadDart];
+}
+
+- (sky::shell::PlatformViewMac*)platformView {
```
Comment (eseidelGoogle) [inline sky/shell/mac/sky_window.mm]: A space somewhere?

File: sky/shell/mac/sky_window.mm [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+  auto view = static_cast<sky::shell::PlatformViewMac*>(_shell_view->view());
+  DCHECK(view);
+  return view;
+}
+
+#pragma mark - Responder overrides
+
+- (void)dispatchEvent:(NSEvent *)event phase:(NSEventPhase) phase {
```
Comment (eseidelGoogle) [inline sky/shell/mac/sky_window.mm]: ) phase?

**Comment by eseidelGoogle** [comment]
lgtm, but I think we could go a couple more rounds here.  It's also definitely not clear to me that you want to be an NSWindow, sublcassing NSWindow is typically wrong.  You probably just want to have a normal NSWindow and be the contentView on it?

**Comment by chinmaygarde** [comment]
Will get rid of the window subclass. Also, will try to figure out how I can get the old hooks with the formatter and linter working.

**Pull request merged**